clipboard_linux.c

```diff
@@ -11,36 +11,75 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <string.h>
+#include <dlfcn.h>
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 
 // syncStatus is a function from the Go side.
 extern void syncStatus(uintptr_t handle, int status);
 
 int clipboard_test() {
+	void *libX11 = dlopen("libX11.so", RTLD_LAZY);
+	if (!libX11) {
+		return -1;
+	}
+
+	Display* (*XOpenDisplay)(int);
+	void (*XCloseDisplay)(Display*);
+
+	XOpenDisplay = (Display* (*)(int)) dlsym(libX11, "XOpenDisplay");
+	XCloseDisplay = (void (*)(Display*)) dlsym(libX11, "XCloseDisplay");
+
     Display* d = NULL;
     for (int i = 0; i < 42; i++) {
-        d = XOpenDisplay(0);
+        d = (*XOpenDisplay)(0);
         if (d == NULL) {
             continue;
         }
         break;
     }
-
     if (d == NULL) {
         return -1;
     }
-    XCloseDisplay(d);
+    (*XCloseDisplay)(d);
+	dlclose(libX11);
     return 0;
 }
 
 // clipboard_write writes the given buf of size n as type typ.
 // if start is provided, the value of start will be changed to 1 to indicate
 // if the write is availiable for reading.
 int clipboard_write(char *typ, unsigned char *buf, size_t n, uintptr_t handle) {
+	void *libX11 = dlopen("libX11.so", RTLD_LAZY);
```

Comment:
As I am not entirely familiar with dlopen, I wonder if there is a lot more costly when we load the .so at every call?

Reply:
You're right, I'll try to rewrite it to that just `clipboard_test` will open up everything.
But then I don't know where to call `dlclose`

Reply:
I see. What's consequence if dlclose is not called in the whole runtime?

the libX11 pointer might be wrapped in a Go pointer, then being allocated and reopened using sync.Pool. The allocate process sets a runtime finalizer to that Go pointer. As long as no one uses, meaning that after two GC cycle, the finalizer can call dlclose.

Reply:
I'm not 100% sure, but found some code https://github.com/rajrakeshdr/linux_core/blob/9252cba42283e71d7858292a400f6c94043ee8ff/core/src/runtime.c that doesn't call dlclose, only when there is a error in loading functions.
We need the library to be loaded at all times and exiting the process will automatically close the lib.